File: blf_lib/src/blam/halo3/release/saved_games/scenario_map_variant.rs
use binrw::{BinRead, BinWrite};
use serde::{Deserialize, Serialize};
use blf_lib::io::bitstream::{c_bitstream_reader, c_bitstream_writer};
use blf_lib::TEST_BIT;
use crate::blam::common::math::real_math::{real_point3d, real_rectangle3d};
use crate::blam::halo3::release::saved_games::saved_game_files::s_content_item_metadata;
use blf_lib::types::array::StaticArray;
use crate::blam::common::math::real_math::real_vector3d;
use crate::blam::common::simulation::simulation_encoding::{simulation_read_quantized_position, simulation_write_quantized_position};
use serde_hex::{SerHex,StrictCap};
use blf_lib_derive::TestSize;
use blf_lib_derivable::result::BLFLibResult;
use crate::types::bool::Bool;
use crate::types::numbers::Float32;

const k_object_type_count: usize = 14;

#[derive(Default, PartialEq, Debug, Clone, Serialize, Deserialize, BinRead, BinWrite, TestSize)]
#[Size(0xE090)]
pub struct c_map_variant {
    pub m_metadata: s_content_item_metadata,
    pub m_map_variant_version: u16,
    pub m_number_of_scenario_objects: u16,
    pub m_number_of_variant_objects: u16,
    pub m_number_of_placeable_object_quotas: u16,
    pub m_map_id: u32,
    pub m_world_bounds: real_rectangle3d,
    pub m_game_engine_subtype: u32,
    pub m_maximum_budget: Float32,
    pub m_spent_budget: Float32,
    pub m_helpers_enabled: Bool,
    pub m_built_in: Bool,
    // #[serde(skip_serializing,skip_deserializing)]
    // __pad12A: [u8; 2],
    #[serde(with = "SerHex::<StrictCap>")]
    #[brw(align_before = 4)]
    pub m_original_map_rsa_signature_hash: u32,
    pub m_variant_objects: StaticArray<s_variant_object_datum, 640>, // 0x130
    pub m_object_type_start_index: StaticArray<i16, k_object_type_count>, // 0xD330 // Correct
    pub m_quotas: StaticArray<s_variant_quota, 256>,
    #[brw(pad_after = 4)]
    pub m_gamestate_indices: StaticArray<i32, 80>,
}

impl c_map_variant {
    pub fn encode(&self, bitstream: &mut c_bitstream_writer) -> BLFLibResult {
        self.m_metadata.encode(bitstream)?;
        bitstream.write_integer(self.m_map_variant_version as u32, 8)?;
        bitstream.write_integer(self.m_original_map_rsa_signature_hash, 32)?;
        bitstream.write_integer(self.m_number_of_scenario_objects as u32, 10)?;
        bitstream.write_integer(self.m_number_of_variant_objects as u32, 10)?;
        bitstream.write_integer(self.m_number_of_placeable_object_quotas as u32, 9)?;
        bitstream.write_integer(self.m_map_id, 32)?;
        bitstream.write_bool(self.m_built_in)?;
        bitstream.write_raw(self.m_world_bounds, 0xC0)?;
        bitstream.write_integer(self.m_game_engine_subtype, 4)?;
        bitstream.write_float(self.m_maximum_budget, 32)?;
        bitstream.write_float(self.m_spent_budget, 32)?;

        for i in 0..self.m_number_of_variant_objects as usize {
            let variant_object = self.m_variant_objects[i];

            if variant_object.flags & 0x3FF == 0 // 0x3FF is 10 bits, there's 10 flags. If none are set...
            {
                bitstream.write_bool(false)?; // variant_object_exists
            }
            else
            {
                bitstream.write_bool(true)?; // variant_object_exists
                bitstream.write_integer(variant_object.flags as u32, 16)?;
                bitstream.write_integer(variant_object.variant_quota_index as u32, 32)?;

                if TEST_BIT!(variant_object.flags, 8) // spawns relative
                {
                    bitstream.write_bool(true)?; // parent-object-exists
                    bitstream.write_raw(variant_object.parent_object_identifier, 64)?;
                }
                else
                {
                    bitstream.write_bool(false)?; // parent-object-exists
                }

                if !TEST_BIT!(variant_object.flags, 1) && i < self.m_number_of_scenario_objects as usize  //edited
                {
                    bitstream.write_bool(false)?;
                }
                else
                {
                    bitstream.write_bool(true)?;
                    simulation_write_quantized_position(bitstream, &variant_object.position, 16, false, &self.m_world_bounds)?;
                    bitstream.write_axes(&variant_object.forward, &variant_object.up)?;
                    bitstream.write_integer(variant_object.multiplayer_game_object_properties.object_type as u32, 8)?;
                    bitstream.write_integer(variant_object.multiplayer_game_object_properties.symmetry_placement_flags as u32, 8)?;
                    bitstream.write_integer(variant_object.multiplayer_game_object_properties.game_engine_flags as u32, 16)?;
                    bitstream.write_integer(variant_object.multiplayer_game_object_properties.shared_storage as u32, 8)?;
                    bitstream.write_integer(variant_object.multiplayer_game_object_properties.spawn_time as u32, 8)?;
                    bitstream.write_integer(variant_object.multiplayer_game_object_properties.owner_team as u32, 8)?;
                    bitstream.write_integer(variant_object.multiplayer_game_object_properties.boundary_shape as u32, 8)?;

                    match variant_object.multiplayer_game_object_properties.boundary_shape {
                        1 => { // sphere
                            bitstream.write_quantized_real(variant_object.multiplayer_game_object_properties.boundary_size, 0.0, 60.0, 16, false, false)?;
                            bitstream.write_quantized_real(variant_object.multiplayer_game_object_properties.boundary_negative_height, 0.0, 60.0, 16, false, false)?;
                        }
                        2 => { // cylinder
                            bitstream.write_quantized_real(variant_object.multiplayer_game_object_properties.boundary_size, 0.0, 60.0, 16, false, false)?;
                            bitstream.write_quantized_real(variant_object.multiplayer_game_object_properties.boundary_box_length, 0.0, 60.0, 16, false, false)?;
                            bitstream.write_quantized_real(variant_object.multiplayer_game_object_properties.boundary_positive_height, 0.0, 60.0, 16, false, false)?;
                        }
                        3 => { // box
                            bitstream.write_quantized_real(variant_object.multiplayer_game_object_properties.boundary_size, 0.0, 60.0, 16, false, false)?;
                            bitstream.write_quantized_real(variant_object.multiplayer_game_object_properties.boundary_box_length, 0.0, 60.0, 16, false, false)?;
                            bitstream.write_quantized_real(variant_object.multiplayer_game_object_properties.boundary_positive_height, 0.0, 60.0, 16, false, false)?;
                            bitstream.write_quantized_real(variant_object.multiplayer_game_object_properties.boundary_negative_height, 0.0, 60.0, 16, false, false)?;
                        }
                        _ => { }
                    }
                }
            }
        }

        for i in 0..k_object_type_count {
            bitstream.write_integer((self.m_object_type_start_index[i] + 1) as u32, 9)?;
        }

        for i in 0..self.m_number_of_placeable_object_quotas as usize {
            let object_quota = self.m_quotas[i];
            bitstream.write_integer(object_quota.object_definition_index, 32)?;
            bitstream.write_integer(object_quota.minimum_count as u32, 8)?;
            bitstream.write_integer(object_quota.maximum_count as u32, 8)?;
            bitstream.write_integer(object_quota.placed_on_map as u32, 8)?;
            bitstream.write_integer(object_quota.maximum_allowed as u32, 8)?;
            bitstream.write_float(object_quota.price_per_item, 32)?;
        }

        Ok(())
    }

    pub fn decode(&mut self, bitstream: &mut c_bitstream_reader) -> BLFLibResult {
        self.m_metadata.decode(bitstream)?;
        self.m_map_variant_version = bitstream.read_integer(8)? as u16;
        self.m_original_map_rsa_signature_hash = bitstream.read_integer(32)?;
        self.m_number_of_scenario_objects = bitstream.read_u16(10)?;
        self.m_number_of_variant_objects = bitstream.read_u16(10)?;
        self.m_number_of_placeable_object_quotas = bitstream.read_u16(9)?;
        self.m_map_id = bitstream.read_integer(32)?;
        self.m_built_in = Bool::from(bitstream.read_bool()?);
        self.m_world_bounds = bitstream.read_raw(0xC0)?;
        self.m_game_engine_subtype = bitstream.read_integer(4)?;
        self.m_maximum_budget = bitstream.read_float(32)?;
        self.m_spent_budget = bitstream.read_float(32)?;

        for i in 0..self.m_number_of_variant_objects as usize {
            let variant_object = &mut self.m_variant_objects.get_mut()[i];

            let variant_object_exists = bitstream.read_bool()?;

            if !variant_object_exists {
                continue;
            }

            variant_object.flags = bitstream.read_u16(16)?;
            variant_object.variant_quota_index = bitstream.read_signed_integer(32)?;

            let parent_object_exists = bitstream.read_bool()?;
            if parent_object_exists {
                variant_object.parent_object_identifier = bitstream.read_raw(64)?;
            }

            let position_exists = bitstream.read_bool()?;

            if !position_exists {
                continue;
            }

            simulation_read_quantized_position(bitstream, &mut variant_object.position, 16, &self.m_world_bounds)?;
            bitstream.read_axis::<8, 19>(&mut variant_object.forward, &mut variant_object.up)?;
            variant_object.multiplayer_game_object_properties.object_type = bitstream.read_signed_integer(8)? as i8;
            variant_object.multiplayer_game_object_properties.symmetry_placement_flags = bitstream.read_u8(8)?;
            variant_object.multiplayer_game_object_properties.game_engine_flags = bitstream.read_u16(16)?;
            variant_object.multiplayer_game_object_properties.shared_storage = bitstream.read_u8(8)?;
            variant_object.multiplayer_game_object_properties.spawn_time = bitstream.read_u8(8)?;
            variant_object.multiplayer_game_object_properties.owner_team = bitstream.read_u8(8)? as i8;
            variant_object.multiplayer_game_object_properties.boundary_shape = bitstream.read_u8(8)?;

            match variant_object.multiplayer_game_object_properties.boundary_shape {
                1 => { // sphere
                    variant_object.multiplayer_game_object_properties.boundary_size = bitstream.read_quantized_real(0.0, 60.0, 16, false, false)?;
                    variant_object.multiplayer_game_object_properties.boundary_negative_height = bitstream.read_quantized_real(0.0, 60.0, 16, false, false)?;
                }
                2 => { // cylinder
                    variant_object.multiplayer_game_object_properties.boundary_size = bitstream.read_quantized_real(0.0, 60.0, 16, false, false)?;
                    variant_object.multiplayer_game_object_properties.boundary_box_length = bitstream.read_quantized_real(0.0, 60.0, 16, false, false)?;
                    variant_object.multiplayer_game_object_properties.boundary_positive_height = bitstream.read_quantized_real(0.0, 60.0, 16, false, false)?;
                }
                3 => { // box
                    variant_object.multiplayer_game_object_properties.boundary_size = bitstream.read_quantized_real(0.0, 60.0, 16, false, false)?;
                    variant_object.multiplayer_game_object_properties.boundary_box_length = bitstream.read_quantized_real(0.0, 60.0, 16, false, false)?;
                    variant_object.multiplayer_game_object_properties.boundary_positive_height = bitstream.read_quantized_real(0.0, 60.0, 16, false, false)?;
                    variant_object.multiplayer_game_object_properties.boundary_negative_height = bitstream.read_quantized_real(0.0, 60.0, 16, false, false)?;
                }
                _ => { }
            }
        }

        for i in 0..k_object_type_count {
            self.m_object_type_start_index.get_mut()[i] = bitstream.read_integer(9)? as i16 - 1;
        }

        for i in 0..self.m_number_of_placeable_object_quotas as usize {
            let object_quota = &mut self.m_quotas.get_mut()[i];
            object_quota.object_definition_index = bitstream.read_integer(32)?;
            object_quota.minimum_count = bitstream.read_integer(8)? as u8;
            object_quota.maximum_count = bitstream.read_integer(8)? as u8;
            object_quota.placed_on_map = bitstream.read_integer(8)? as u8;
            object_quota.maximum_allowed = bitstream.read_integer(8)? as i8;
            object_quota.price_per_item = bitstream.read_float(32)?;
        }

        Ok(())
    }
}

#[derive(Default, PartialEq, Debug, Clone, Copy, Serialize, Deserialize, BinRead, BinWrite)]
pub struct s_variant_quota {
    #[serde(with = "SerHex::<StrictCap>")]
    pub object_definition_index: u32,
    pub minimum_count: u8,
    pub maximum_count: u8,
    pub placed_on_map: u8,
    pub maximum_allowed: i8,
    pub price_per_item: Float32,
}

#[derive(Default, PartialEq, Debug, Clone, Copy, Serialize, Deserialize, BinRead, BinWrite, TestSize)]
#[Size(0x18)]
pub struct s_variant_multiplayer_object_properties_definition {
    pub game_engine_flags: u16,
    pub symmetry_placement_flags: u8, // foo
    pub owner_team: i8, // byte?
    pub shared_storage: u8, // spare_clips, teleporter_channel, spawn_rate
    pub spawn_time: u8,
    pub object_type: i8,
    pub boundary_shape: u8,
    pub boundary_size: Float32, // width or radius
    pub boundary_box_length: Float32,
    pub boundary_positive_height: Float32,
    pub boundary_negative_height: Float32,
}

#[derive(Default, PartialEq, Debug, Clone, Copy, Serialize, Deserialize, BinRead, BinWrite, TestSize)]
#[Size(0x54)]
pub struct s_variant_object_datum {
    pub flags: u16,
    pub reuse_timeout: u16,
    pub object_datum_index: i32,
    pub editor_object_index: i32,
    pub variant_quota_index: i32,
    pub position: real_point3d,
    pub forward: real_vector3d,
    pub up: real_vector3d,
    pub parent_object_identifier: c_object_identifier,
    pub multiplayer_game_object_properties: s_variant_multiplayer_object_properties_definition,
}

#[derive(Default, PartialEq, Debug, Clone, Copy, Serialize, Deserialize, BinRead, BinWrite, TestSize)]
#[Size(0x8)]
pub struct c_object_identifier {
    m_unique_id: i32,
    m_origin_bsp_index: i16,
    m_type: i8,
    m_source: i8,
}

File: blf-ts/src/blam/haloreach/v12065_11_08_24_1738_tu1actual/saved_games/scenario_map_variant.ts
import type {
  c_bitstream_reader,
  c_bitstream_writer,
} from "../../../../bitstream";
import {
  type real_vector3d,
  real_vector3d_default,
} from "../../../../bitstream/math";
import { BlfError } from "../../../../error";
import { AutoMap } from "../../../../helpers/automap";
import type {
  real_point3d,
  real_rectangle3d,
} from "../../../common/math/real_types";
import {
  point_in_rectangle3d,
  real_point3d_default,
  real_rectangle3d_default,
} from "../../../common/math/real_types";
import { c_single_language_string_table } from "../game/string_table";
import { read_axes, write_axes } from "../memory/bitstream_extensions";
import {
  simulation_read_position,
  simulation_write_position,
} from "../simulation/simulation_encoding";
import {
  content_item_metadata_decode,
  content_item_metadata_encode,
  s_content_item_display_metadata,
  s_content_item_general_metadata,
  s_content_item_history,
  s_content_item_metadata,
} from "./saved_game_files";

export const k_maximum_variant_objects = 651;
export const k_maximum_variant_quotas = 256;

/** Matches `e_boundary_shape` in blf_lib `scenario_map_variant.rs`. */
export enum e_boundary_shape {
  unused = 0,
  sphere = 1,
  cylinder = 2,
  box = 3,
}

export class s_variant_quota {
  @AutoMap(() => Number)
  minimum_count = 0;
  @AutoMap(() => Number)
  maximum_count = 0;
  @AutoMap(() => Number)
  placed_on_map = 0;

  decode(bitstream: c_bitstream_reader): void {
    this.minimum_count = bitstream.read_integer("minimum-count", 8);
    this.maximum_count = bitstream.read_integer("maximum-count", 8);
    this.placed_on_map = bitstream.read_integer("placed-on-map", 8);
  }

  encode(bitstream: c_bitstream_writer): void {
    bitstream.write_integer(this.minimum_count, 8);
    bitstream.write_integer(this.maximum_count, 8);
    bitstream.write_integer(this.placed_on_map, 8);
  }
}

export class s_multiplayer_object_boundary {
  @AutoMap(() => e_boundary_shape)
  shape: e_boundary_shape = e_boundary_shape.unused;
  @AutoMap(() => Number)
  size = 0;
  @AutoMap(() => Number)
  box_length = 0;
  @AutoMap(() => Number)
  positive_height = 0;
  @AutoMap(() => Number)
  negative_height = 0;

  static decode(
    bitstream: c_bitstream_reader
  ): s_multiplayer_object_boundary | null {
    const boundary = new s_multiplayer_object_boundary();
    boundary.shape = bitstream.read_enum("shape", 2, e_boundary_shape);

    switch (boundary.shape) {
      case e_boundary_shape.unused:
        return null;
      case e_boundary_shape.sphere:
        boundary.size = bitstream.read_quantized_real(0, 200, 11, false, true);
        break;
      case e_boundary_shape.cylinder:
        boundary.size = bitstream.read_quantized_real(0, 200, 11, false, true);
        boundary.positive_height = bitstream.read_quantized_real(
          0,
          200,
          11,
          false,
          true
        );
        boundary.negative_height = bitstream.read_quantized_real(
          0,
          200,
          11,
          false,
          true
        );
        break;
      case e_boundary_shape.box:
        boundary.size = bitstream.read_quantized_real(0, 200, 11, false, true);
        boundary.box_length = bitstream.read_quantized_real(
          0,
          200,
          11,
          false,
          true
        );
        boundary.positive_height = bitstream.read_quantized_real(
          0,
          200,
          11,
          false,
          true
        );
        boundary.negative_height = bitstream.read_quantized_real(
          0,
          200,
          11,
          false,
          true
        );
        break;
    }

    return boundary;
  }

  encode(bitstream: c_bitstream_writer): void {
    bitstream.write_enum(this.shape, 2, e_boundary_shape);

    switch (this.shape) {
      case e_boundary_shape.unused:
        break;
      case e_boundary_shape.sphere:
        bitstream.write_quantized_real(this.size, 0, 200, 11, false, true);
        break;
      case e_boundary_shape.cylinder:
        bitstream.write_quantized_real(this.size, 0, 200, 11, false, true);
        bitstream.write_quantized_real(
          this.positive_height,
          0,
          200,
          11,
          false,
          true
        );
        bitstream.write_quantized_real(
          this.negative_height,
          0,
          200,
          11,
          false,
          true
        );
        break;
      case e_boundary_shape.box:
        bitstream.write_quantized_real(this.size, 0, 200, 11, false, true);
        bitstream.write_quantized_real(
          this.box_length,
          0,
          200,
          11,
          false,
          true
        );
        bitstream.write_quantized_real(
          this.positive_height,
          0,
          200,
          11,
          false,
          true
        );
        bitstream.write_quantized_real(
          this.negative_height,
          0,
          200,
          11,
          false,
          true
        );
        break;
    }
  }
}

export class s_variant_multiplayer_object_properties_definition_location_data {
  @AutoMap(() => Number)
  location_name_index = 0;
}

export class s_variant_multiplayer_object_properties_definition_teleporter_data {
  @AutoMap(() => Number)
  channel = 0;
  @AutoMap(() => Number)
  passability = 0;
}

export class s_variant_multiplayer_object_properties_definition_weapon_data {
  @AutoMap(() => Number)
  spare_clips = 0;
}

export class s_variant_multiplayer_object_properties_definition {
  @AutoMap(() => s_multiplayer_object_boundary)
  boundary?: s_multiplayer_object_boundary;
  @AutoMap(() => Number)
  game_engine_flags = 0;
  @AutoMap(() => Number)
  user_data = 0;
  @AutoMap(() => Number)
  spawn_time = 0;
  @AutoMap(() => Number)
  cached_type = 0;
  @AutoMap(() => Number)
  label_index = 0;
  @AutoMap(() => Number)
  placement_flags = 0;
  @AutoMap(() => Number)
  team = 0;
  @AutoMap(() => Number)
  primary_change_color_index = 0;
  @AutoMap(
    () => s_variant_multiplayer_object_properties_definition_location_data
  )
  location_data?: s_variant_multiplayer_object_properties_definition_location_data;
  @AutoMap(
    () => s_variant_multiplayer_object_properties_definition_teleporter_data
  )
  teleporter_data?: s_variant_multiplayer_object_properties_definition_teleporter_data;
  @AutoMap(() => s_variant_multiplayer_object_properties_definition_weapon_data)
  weapon_data?: s_variant_multiplayer_object_properties_definition_weapon_data;

  decode(bitstream: c_bitstream_reader): void {
    this.boundary =
      s_multiplayer_object_boundary.decode(bitstream) ?? undefined;
    this.user_data = bitstream.read_integer("user-data", 8);
    this.spawn_time = bitstream.read_integer("spawn-time", 8);
    this.cached_type = bitstream.read_integer("cached-type", 5);
    this.label_index = bitstream.read_index("label-index", 256, 8);
    this.placement_flags = bitstream.read_integer("placement-flags", 8);
    this.team = bitstream.read_integer("team", 4) - 1;
    this.primary_change_color_index = bitstream.read_index(
      "primary-change-color-index",
      8,
      3
    );

    this.location_data = undefined;
    this.teleporter_data = undefined;
    this.weapon_data = undefined;

    switch (this.cached_type) {
      case 1:
        this.weapon_data =
          new s_variant_multiplayer_object_properties_definition_weapon_data();
        this.weapon_data.spare_clips = bitstream.read_integer("spare-clips", 8);
        break;
      case 12:
      case 13:
      case 14:
        this.teleporter_data =
          new s_variant_multiplayer_object_properties_definition_teleporter_data();
        this.teleporter_data.channel = bitstream.read_integer("channel", 5);
        this.teleporter_data.passability = bitstream.read_integer(
          "passability",
          5
        );
        break;
      case 19:
        this.location_data =
          new s_variant_multiplayer_object_properties_definition_location_data();
        this.location_data.location_name_index = bitstream.read_index(
          "location-name-index",
          255,
          8
        );
        break;
    }
  }

  encode(bitstream: c_bitstream_writer): void {
    (this.boundary ?? new s_multiplayer_object_boundary()).encode(bitstream);
    bitstream.write_integer(this.user_data, 8);
    bitstream.write_integer(this.spawn_time, 8);
    bitstream.write_integer(this.cached_type, 5);
    bitstream.write_index(this.label_index, 256, 8);
    bitstream.write_integer(this.placement_flags, 8);
    bitstream.write_integer(this.team + 1, 4);
    bitstream.write_index(this.primary_change_color_index, 8, 3);

    switch (this.cached_type) {
      case 1:
        if (!this.weapon_data) {
          throw new BlfError(
            "Tried to encode a weapon with no weapon data provided."
          );
        }
        bitstream.write_integer(this.weapon_data.spare_clips, 8);
        break;
      case 12:
      case 13:
      case 14:
        if (!this.teleporter_data) {
          throw new BlfError(
            "Tried to encode a teleporter with no teleporter data provided."
          );
        }
        bitstream.write_integer(this.teleporter_data.channel, 5);
        bitstream.write_integer(this.teleporter_data.passability, 5);
        break;
      case 19:
        if (!this.location_data) {
          throw new BlfError(
            "Tried to encode a location name with no name provided."
          );
        }
        bitstream.write_index(this.location_data.location_name_index, 255, 8);
        break;
    }
  }
}

export class s_variant_object_datum {
  @AutoMap(() => Number)
  flags = 0;
  @AutoMap(() => Number)
  reuse_timeout = 0;
  @AutoMap(() => Number)
  object_datum_index = 0;
  @AutoMap(() => Number)
  editor_object_index = 0;
  @AutoMap(() => Number)
  variant_quota_index = 0;
  @AutoMap(() => Number)
  variant_index = 0;
  @AutoMap(() => Object)
  position: real_point3d = real_point3d_default();
  @AutoMap(() => Object)
  forward: real_vector3d = real_vector3d_default();
  @AutoMap(() => Object)
  up: real_vector3d = real_vector3d_default();
  @AutoMap(() => Number)
  spawn_relative_to = 0;
  @AutoMap(() => s_variant_multiplayer_object_properties_definition)
  multiplayer_game_object_properties =
    new s_variant_multiplayer_object_properties_definition();

  decode(bitstream: c_bitstream_reader, world_bounds: real_rectangle3d): void {
    if (bitstream.read_bool("variant_object_exists")) {
      this.flags = bitstream.read_integer("flags", 2);
      this.variant_quota_index = bitstream.read_index(
        "variant-quota-index",
        k_maximum_variant_quotas,
        8
      );
      this.variant_index = bitstream.read_index("variant-index", 32, 5);
      simulation_read_position(bitstream, this.position, 21, world_bounds);
      read_axes(bitstream, this.forward, this.up, 14, 20);
      this.spawn_relative_to =
        bitstream.read_integer("spawn-relative-to", 10) - 1;
      this.multiplayer_game_object_properties.decode(bitstream);
    }
  }

  encode(bitstream: c_bitstream_writer, world_bounds: real_rectangle3d): void {
    if ((this.flags & 0x3ff) === 0) {
      bitstream.write_bool(false);
      return;
    }

    bitstream.write_bool(true);
    bitstream.write_integer(this.flags, 2);
    bitstream.write_index(
      this.variant_quota_index,
      k_maximum_variant_quotas,
      8
    );
    bitstream.write_index(this.variant_index, 32, 5);
    simulation_write_position(bitstream, this.position, 21, world_bounds);
    write_axes(bitstream, this.forward, this.up, 14, 20);
    bitstream.write_integer(this.spawn_relative_to + 1, 10);
    this.multiplayer_game_object_properties.encode(bitstream);
  }
}

export class c_object_identifier {
  m_unique_id = 0;
  m_origin_bsp_index = 0;
  m_type = 0;
  m_source = 0;
}

function read_real_rectangle3d(
  bitstream: c_bitstream_reader
): real_rectangle3d {
  const raw = bitstream.read_raw_data(0xc0);
  const view = new DataView(raw.buffer, raw.byteOffset, raw.byteLength);
  return {
    x: { lower: view.getFloat32(0, false), upper: view.getFloat32(4, false) },
    y: { lower: view.getFloat32(8, false), upper: view.getFloat32(12, false) },
    z: { lower: view.getFloat32(16, false), upper: view.getFloat32(20, false) },
  };
}

function write_real_rectangle3d(
  bitstream: c_bitstream_writer,
  bounds: real_rectangle3d
): void {
  const raw = new Uint8Array(24);
  const view = new DataView(raw.buffer);
  view.setFloat32(0, bounds.x.lower, false);
  view.setFloat32(4, bounds.x.upper, false);
  view.setFloat32(8, bounds.y.lower, false);
  view.setFloat32(12, bounds.y.upper, false);
  view.setFloat32(16, bounds.z.lower, false);
  view.setFloat32(20, bounds.z.upper, false);
  bitstream.write_raw_data(raw, 0xc0);
}

/**
 * Reach map variant packed body.
 * Mirrors `c_map_variant` in blf_lib — not a direct memory layout.
 */
export class c_map_variant {
  @AutoMap(() => s_content_item_metadata)
  m_metadata = (() => {
    const metadata = new s_content_item_metadata();
    metadata.general = new s_content_item_general_metadata();
    metadata.display = new s_content_item_display_metadata();
    metadata.creation_history = new s_content_item_history();
    metadata.modification_history = new s_content_item_history();
    return metadata;
  })();
  @AutoMap(() => Number)
  m_map_variant_version = 0;
  @AutoMap(() => Number)
  m_number_of_placeable_object_quotas = 0;
  @AutoMap(() => Number)
  m_map_id = 0;
  @AutoMap(() => Object)
  m_world_bounds: real_rectangle3d = real_rectangle3d_default();
  @AutoMap(() => Number)
  m_maximum_budget = 0;
  @AutoMap(() => Number)
  m_spent_budget = 0;
  @AutoMap(() => Boolean)
  m_helpers_enabled = false;
  @AutoMap(() => Boolean)
  m_built_in = false;
  @AutoMap(() => Boolean)
  m_built_from_xml = false;
  @AutoMap(() => Number)
  m_original_map_rsa_signature_hash = 0;
  @AutoMap(() => Number)
  m_scenario_palette_crc = 0;
  @AutoMap(() => c_single_language_string_table)
  m_string_table = new c_single_language_string_table(256, 4096, 12, 13, 9);
  @AutoMap(() => [s_variant_object_datum])
  m_variant_objects: s_variant_object_datum[] = Array.from(
    { length: k_maximum_variant_objects },
    () => new s_variant_object_datum()
  );
  @AutoMap(() => [s_variant_quota])
  m_quotas: s_variant_quota[] = Array.from(
    { length: k_maximum_variant_quotas },
    () => new s_variant_quota()
  );

  decode(bitstream: c_bitstream_reader): void {
    content_item_metadata_decode(bitstream, this.m_metadata);
    this.m_map_variant_version = bitstream.read_integer(
      "map-variant-version",
      8
    );
    this.m_original_map_rsa_signature_hash = bitstream.read_integer(
      "original-map-rsa-signature-hash",
      32
    );
    this.m_scenario_palette_crc = bitstream.read_integer(
      "scenario-palette-crc",
      32
    );
    this.m_number_of_placeable_object_quotas = bitstream.read_integer(
      "number-of-placeable-object-quotas",
      9
    );
    this.m_map_id = bitstream.read_integer("map-id", 32);
    this.m_built_in = bitstream.read_bool("built-in");
    this.m_built_from_xml = bitstream.read_bool("built-from-xml");
    this.m_world_bounds = read_real_rectangle3d(bitstream);
    this.m_maximum_budget = bitstream.read_integer("maximum-budget", 32);
    this.m_spent_budget = bitstream.read_integer("spent-budget", 32);
    this.m_string_table.decode(bitstream);

    for (let i = 0; i < k_maximum_variant_objects; i++) {
      this.m_variant_objects[i]!.decode(bitstream, this.m_world_bounds);
    }

    const quota_count = Math.min(
      k_maximum_variant_quotas,
      this.m_number_of_placeable_object_quotas
    );
    for (let i = 0; i < quota_count; i++) {
      this.m_quotas[i]!.decode(bitstream);
    }
  }

  encode(bitstream: c_bitstream_writer): void {
    content_item_metadata_encode(bitstream, this.m_metadata);
    bitstream.write_integer(this.m_map_variant_version, 8);
    bitstream.write_integer(this.m_original_map_rsa_signature_hash, 32);
    bitstream.write_integer(this.m_scenario_palette_crc, 32);
    bitstream.write_integer(this.m_number_of_placeable_object_quotas, 9);
    bitstream.write_integer(this.m_map_id, 32);
    bitstream.write_bool(this.m_built_in);
    bitstream.write_bool(this.m_built_from_xml);
    write_real_rectangle3d(bitstream, this.m_world_bounds);
    bitstream.write_integer(this.m_maximum_budget, 32);
    bitstream.write_integer(this.m_spent_budget, 32);
    this.m_string_table.encode(bitstream);

    for (let i = 0; i < k_maximum_variant_objects; i++) {
      this.m_variant_objects[i]!.encode(bitstream, this.m_world_bounds);
    }

    const quota_count = Math.min(
      k_maximum_variant_quotas,
      this.m_number_of_placeable_object_quotas
    );
    for (let i = 0; i < quota_count; i++) {
      this.m_quotas[i]!.encode(bitstream);
    }
  }
}

export function create_default_map_variant(): c_map_variant {
  return new c_map_variant();
}

/** Convenience helper used by tests and tooling. */
export function map_variant_point_in_bounds(
  map_variant: c_map_variant,
  position: real_point3d
): boolean {
  return point_in_rectangle3d(position, map_variant.m_world_bounds);
}
